Source/Dafny/DafnyAst.cs

```diff
@@ -9039,27 +9243,75 @@ public class MemberSelectExpr : Expression {
+
+    /// If Member is a Function or Method, then TypeApplication is the list of type arguments used with the receiver
+    /// type and the function/method itself. If it is a Field, then TypeApplication is the list of type arguments used
+    /// with the receiver type. The receiver type may be different than the enclosing class of the member; this will
+    /// happen if the member is declared in a trait and the receiver is of some type that extends that trait.
+    /// However, for a static member, the said receiver type is always the enclosing class of the member, even if
+    /// the type of an explicitly given receiver is some subclass thereof.
+    public List<Type> TypeApplication;
 
     public Dictionary<TypeParameter, Type> TypeArgumentSubstitutions() {
       Contract.Requires(WasResolved());
       Contract.Ensures(Contract.Result<Dictionary<TypeParameter, Type>>() != null);
       Contract.Ensures(Contract.Result<Dictionary<TypeParameter, Type>>().Count == TypeApplication.Count);
 
-      var icallable = Member as ICallable;
-      Contract.Assert(Member.EnclosingClass.TypeArgs.Count + (icallable == null ? 0 : icallable.TypeArgs.Count) == TypeApplication.Count);  // a consequence of proper resolution
       var subst = new Dictionary<TypeParameter, Type>();
       var i = 0;
       foreach (var tp in Member.EnclosingClass.TypeArgs) {
         subst.Add(tp, TypeApplication[i]);
         i++;
       }
+      if (Member is ICallable icallable) {
+        foreach (var tp in icallable.TypeArgs) {
+          subst.Add(tp, TypeApplication[i]);
+          i++;
+        }
+      }
+      return subst;
+    }
+
+    public Dictionary<TypeParameter, Type> TypeArgumentSubstitutionsWithParents() {
+      Contract.Requires(WasResolved());
+      Contract.Ensures(Contract.Result<Dictionary<TypeParameter, Type>>() != null);
+
+      // Determine the class
+      TopLevelDeclWithMembers cl;
+      // Expand the type down to its non-null type, if any
+      var receiverType = Obj.Type.AsNonNullRefType;
+      if (receiverType != null) {
+        cl = (TopLevelDeclWithMembers)((NonNullTypeDecl)receiverType.ResolvedClass).ViewAsClass;  // TODO: why is this ever any different than what the else branch does?
+      } else {
+        // Expand type all the way
+        receiverType = Obj.Type.NormalizeExpand() as UserDefinedType;
+        cl = receiverType?.ResolvedClass as TopLevelDeclWithMembers;
+      }
+      if (cl == null) {
+        return TypeArgumentSubstitutions();  // TODO: do we ever get here? (it seems Obj.Type must be a value type)
+      }
+
+      // build map from class+member formal type parameters to types
+      var icallable = Member as ICallable;
+      var subst = new Dictionary<TypeParameter, Type>();
+      var i = 0;
+      foreach (var tp in cl.TypeArgs) {
+        subst.Add(tp, TypeApplication[i]);
```

Comment:
We can get an Index out of bounds if we have a class such as myClass\<T, R\> extends myTrait\<T\>. SImplest repro I can come up with:
```
module TraitBug {
    trait MyTrait<T> {
        ghost const Repr: set<object>;
    }

    class MyClass<T, R> extends MyTrait<T> {
        constructor() {
            this.Repr := {this};
        }
    }

    class OtherClass {
        ghost const Repr: set<object>;
        var foo: MyTrait<int>;

        constructor()
        {
            var foo := new MyClass<int, int>();
            this.foo := foo;
            this.Repr := {this} + foo.Repr; // This is where it runs into trouble
        }
    }

    method someMethod() {
        var foo := new OtherClass();
    }
}
```

Reply:
I've distilled this example into `module ReceiverResolution` in TraitVerify.dfy.